tower-http/src/cors.rs

```diff
@@ -0,0 +1,798 @@
+    /// use http::{HeaderValue, request::Parts};
+    ///
+    /// let layer = CorsLayer::new().allow_origin(
+    ///     Origin::predicate(|origin: &HeaderValue, _request_head: &Parts| {
+    ///         origin.as_bytes().ends_with(b".rust-lang.org")
+    ///     })
+    /// );
+    /// ```
+    ///
+    /// Note that multiple calls to this method will override any previous
+    /// calls.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin
+    pub fn allow_origin<T>(mut self, origin: T) -> Self
+    where
+        T: Into<AnyOr<Origin>>,
+    {
+        self.allow_origin = Some(origin.into());
+        self
+    }
+
+    /// Set the value of the [`Access-Control-Expose-Headers`][mdn] header.
+    ///
+    /// ```
+    /// use tower_http::cors::CorsLayer;
+    /// use http::header::CONTENT_ENCODING;
+    ///
+    /// let layer = CorsLayer::new().expose_headers(vec![CONTENT_ENCODING]);
+    /// ```
+    ///
+    /// All headers can be allowed with
+    ///
+    /// ```
+    /// use tower_http::cors::{CorsLayer, Any};
+    ///
+    /// let layer = CorsLayer::new().expose_headers(Any);
+    /// ```
+    ///
+    /// Note that multiple calls to this method will override any previous
+    /// calls.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Expose-Headers
+    pub fn expose_headers<I>(mut self, headers: I) -> Self
+    where
+        I: Into<AnyOr<Vec<HeaderName>>>,
+    {
+        self.expose_headers = Some(match headers.into().0 {
+            AnyOrInner::Any => WILDCARD.parse().unwrap(),
+            AnyOrInner::Value(headers) => separated_by_commas(headers),
+        });
+        self
+    }
+}
+
+/// Represents a wildcard value (`*`) used with some CORS headers such as
+/// [`CorsLayer::allow_methods`].
+#[derive(Debug, Clone, Copy)]
+pub struct Any;
+
+/// Used to make methods like [`CorsLayer::allow_methods`] more convenient to call.
+///
+/// You shouldn't have to use this type directly.
+#[derive(Debug, Clone, Copy)]
+pub struct AnyOr<T>(AnyOrInner<T>);
+
+#[derive(Debug, Clone, Copy)]
+enum AnyOrInner<T> {
+    Any,
+    Value(T),
+}
+
+impl From<Origin> for AnyOr<Origin> {
+    fn from(origin: Origin) -> Self {
+        AnyOr(AnyOrInner::Value(origin))
+    }
+}
+
+impl<T> From<Any> for AnyOr<T> {
+    fn from(_: Any) -> Self {
+        AnyOr(AnyOrInner::Any)
+    }
+}
+
+impl<I> From<I> for AnyOr<Vec<Method>>
+where
+    I: IntoIterator<Item = Method>,
+{
+    fn from(methods: I) -> Self {
+        AnyOr(AnyOrInner::Value(methods.into_iter().collect()))
+    }
+}
+
+impl<I> From<I> for AnyOr<Vec<HeaderName>>
+where
+    I: IntoIterator<Item = HeaderName>,
+{
+    fn from(headers: I) -> Self {
+        AnyOr(AnyOrInner::Value(headers.into_iter().collect()))
+    }
+}
+
+fn separated_by_commas<I>(into_iter: I) -> HeaderValue
+where
+    I: IntoIterator,
+    I::Item: ToString,
+{
+    let methods = into_iter
```

Comment:
Suggestion to improve: Create mutable bytes buffer and write it directly, then pass it to `HeaderValue::from_maybe_shared` (simplest way is to crete abstraction that implements `fmt::Write` by extending `BytesMut`)

If it is too much of work I can do PR myself later

Reply:
You're right be since these methods are only called during initialization (and not on every `call`) I figured it would be fine.

You're very welcome to submit a follow up PR once this is merge 😊

Reply:
Ah yeah, you're right, it is minor thing, but it would serve as good reference in future as creating `Bytes` directly via formatting is far more efficient than creating temporarily allocations. I'll be sure to make PR